File: core/app/models/course.rb
class Course < ActiveRecord::Base
  belongs_to :subject
  has_many   :listings, dependent: :destroy
  has_one    :latest_listing, -> { latest(true) }, class_name: 'Listing'

  # validates  :number, presence: true, uniqueness: { scope: :subject_id }
  default_scope { order(shortname: :asc) }
  # after_save :send_notification

  before_create { self.uuid ||= SecureRandom.uuid }

  # def send_notification
  #   CoursesResponder.new.call(self)
  # end

  def self.get code, number
    joins(:subject).where("subjects.shortname = ? AND number = ?", shortname, number).first
  end
end

File: packages/yacs-api-client/src/lib/course.ts
import {
  Model,
  SpraypaintBase,
  Attr,
  BelongsTo,
  HasMany
} from "spraypaint";

import { ApplicationRecord } from "./application-record";
import { Subject } from "./subject";
import { Listing } from "./listing";

@Model()
export class Course extends ApplicationRecord {
	static jsonapiType = "courses";

	@Attr() shortname: string;
	@Attr() uuid: string;

	@BelongsTo() subject: Subject;
	@HasMany() listings: Listing[];
}
